tokio-rustls/src/common/mod.rs

```diff
@@ -238,70 +236,49 @@ where
         cx: &mut Context<'_>,
         buf: &mut ReadBuf<'_>,
     ) -> Poll<io::Result<()>> {
-        let prev = buf.remaining();
+        let mut io_pending = false;
 
-        while buf.remaining() != 0 {
-            let mut io_pending = false;
-
-            // read a packet
-            while !self.eof && self.session.wants_read() {
-                match self.read_io(cx) {
-                    Poll::Ready(Ok(0)) => {
-                        self.eof = true;
-                        break;
-                    }
-                    Poll::Ready(Ok(_)) => (),
-                    Poll::Pending => {
-                        io_pending = true;
-                        break;
-                    }
-                    Poll::Ready(Err(err)) => return Poll::Ready(Err(err)),
+        // read a packet
+        while !self.eof && self.session.wants_read() {
+            match self.read_io(cx) {
+                Poll::Ready(Ok(0)) => {
+                    self.eof = true;
+                    break;
                 }
+                Poll::Ready(Ok(_)) => (),
+                Poll::Pending => {
+                    io_pending = true;
+                    break;
+                }
+                Poll::Ready(Err(err)) => return Poll::Ready(Err(err)),
             }
+        }
 
-            return match self.session.reader().read(buf.initialize_unfilled()) {
-                // If Rustls returns `Ok(0)` (while `buf` is non-empty), the peer closed the
-                // connection with a `CloseNotify` message and no more data will be forthcoming.
-                Ok(0) => break,
-
-                // Rustls yielded more data: advance the buffer, then see if more data is coming.
-                Ok(n) => {
-                    buf.advance(n);
-
-                    if self.eof || io_pending {
-                        break;
-                    } else {
-                        continue;
-                    }
-                }
+        match self.session.reader().read(buf.initialize_unfilled()) {
+            // If Rustls returns `Ok(0)` (while `buf` is non-empty), the peer closed the
+            // connection with a `CloseNotify` message and no more data will be forthcoming.
+            //
+            // Rustls yielded more data: advance the buffer, then see if more data is coming.
+            Ok(n) => {
+                buf.advance(n);
```

Comment:
Should we set `self.eof` to `true` if `n == 0 && buf.remaining() > 0`? If `io_pending` is set here, should we not preserve that state for the next call?

Reply:
We don't need to set `self.eof`, because it is only a temporary state. `self.eof` elsewhere is also not needed.

rustls should ensure that it will return 0 only when `CloseNotify` is received, so no additional processing is required.

Reply:
Sorry to drag my feet on this, feel free to just merge if you get enough of my comments, but the comment you added here is sort of superficial in that it only explains what ("no need to modify `self.eof`); what I think would be much more useful is to provide the context like "rustls will only return 0 if is has received `CloseNotify`, in which case no additional processing is required".